src/extensions/metadata.rs

```diff
@@ -0,0 +1,172 @@
+
+use std::fmt::Debug;
+
+use async_std::io::{Read, Write};
+use async_std::prelude::*;
+use async_std::stream::Stream;
+use async_std::sync;
+use imap_proto::types::{MailboxDatum, Metadata};
+use imap_proto::{RequestId, Response};
+
+use crate::client::{validate_str, Session};
+use crate::codec::ResponseData;
+use crate::error::Result;
+use crate::parse::handle_unilateral;
+use crate::types::UnsolicitedResponse;
+
+trait CmdListItemFormat {
+    fn format_as_cmd_list_item(&self) -> String;
+}
+
+impl CmdListItemFormat for Metadata {
+    fn format_as_cmd_list_item(&self) -> String {
+        format!(
+            "{} {}",
+            validate_str(self.entry.as_str()).unwrap(),
+            self.value
+                .as_ref()
+                .map(|v| validate_str(v.as_str()).unwrap())
+                .unwrap_or_else(|| "NIL".to_string())
+        )
+    }
+}
+
+/// Represents variants of DEPTH parameter for GETMETADATA command
+/// See RFC5464 section 4.2.2 "DEPTH GETMETADATA Command Option"
+/// https://tools.ietf.org/html/rfc5464#section-4.2.2
+#[derive(Debug, Copy, Clone)]
+pub enum MetadataDepth {
+    /// Depth 0 for get metadata, no entries below the specified one are returned
+    Zero,
+    /// Depth 1 for get metadata, only entries immediately below the specified one are returned
+    One,
+    /// Depth infinity for get metadata, all entries below the specified one are returned
+    Inf,
+}
+
+impl MetadataDepth {
+    fn depth_str<'a>(self) -> &'a str {
+        match self {
+            MetadataDepth::Zero => "0",
+            MetadataDepth::One => "1",
+            MetadataDepth::Inf => "infinity",
+        }
+    }
+}
+
+async fn parse_metadata<'a, T: Stream<Item = ResponseData> + Unpin>(
+    stream: &'a mut T,
+    unsolicited: sync::Sender<UnsolicitedResponse>,
+    command_tag: RequestId,
+) -> Result<Vec<Metadata>> {
+    let mut res: Vec<Metadata> = Vec::new();
+    while let Some(resp) = stream
+        .take_while(|r| match r.parsed() {
+            Response::Done { tag, .. } => tag != &command_tag,
+            _ => true,
+        })
+        .next()
+        .await
+    {
+        let unsolicited_h = unsolicited.clone();
+        match resp.response {
+            Response::MailboxData(MailboxDatum::MetadataSolicited { mut values, .. }) => {
+                res.append(&mut values);
+            }
+            Response::MailboxData(MailboxDatum::MetadataUnsolicited { mailbox, values }) => {
+                unsolicited
+                    .send(UnsolicitedResponse::Metadata {
+                        mailbox: mailbox.to_string(),
+                        metadata_entries: values.iter().map(|s| (*s).to_string()).collect(),
+                    })
+                    .await;
+            }
+            _ => handle_unilateral(resp, unsolicited_h).await,
+        };
+    }
+    Ok(res)
+}
+
+/// Sends GETMETADATA command to the server and returns the list of entries and their values.
```

Comment:
This should keep a similar interface as other extensions, in that they are methods defined on the `client`, rather than standalone